woocommerce/woocommerce-gateway-stripe
# Fix for Add Payment Method using UPE

# Changes proposed in this Pull Request:

With the changes in #1908 the UPE gateway now inherits from `WC_Gateway_Stripe`. Due to this change, the ability to add a saved payment method from the My Account area was broken. This was due to the `is_available` function in `WC_Gateway_Stripe` returning false since the `saved_cards` property was not set since the parent constructor is never called (on purpose). In addition to this, when fixing that issue it was discovered that the Add payment method screen would still allow credit cards to be saved even if they were not enabled through UPE.

This PR addresses both issues by adding an `is_available` function to `WC_Stripe_UPE_Payment_Gateway`. Within this function it performs an additional check to see if there are any methods enabled in UPE that allow saving a payment method and if none are found return false early. It then calls the parent function which now works after adding and setting the `saved_cards` property in the UPE gateway. The parent function then trickles up its checks accordingly.


# Testing instructions

- With this branch pulled down and UPE enabled with the following settings:
    - Saved cards enabled
    - Only giropay enabled within UPE
- As a logged in customer, go to My Account > Payment Methods and confirm "No saved methods found." message and no ability to add a payment method.
- As the merchant, enable Credit Cards and SEPA within UPE.
- As a logged in customer, go to My Account > Payment Methods and confirm the Add Payment Method button is displayed (along with any previously saved methods being displayed above it).
- Click the Add Payment Method button and confirm UPE allows a CC or SEPA payment method to be added. 
- Repeat this process by adding at least one CC and one SEPA payment method and confirm it saves and is displayed on the My Account > Payment Methods page.
- As the merchant, disable payment via saved cards.
- As a logged in customer, go to My Account > Payment Methods and confirm the Add Payment Method button is NOT displayed.
- As merchant, disable the new checkout experience (UPE), leave payment via saved cards disabled, and leave CC, giropay, and SEPA enabled.
- As a logged in customer, go to My Account > Payment Methods and confirm the Add Payment Method button is still NOT displayed.
- As the merchant, enable payment via saved cards.
- As a logged in customer, go to My Account > Payment Methods and confirm the Add Payment Method button is displayed again.
- Click the Add Payment Method button and confirm the classic gateways are displayed and allow you to create a saved payment method that gets properly displayed on the My Account > Payment Methods.

-------------------
